tensorboard/webapp/widgets/linked_time_fob/types.ts

```diff
@@ -0,0 +1,45 @@
+/* Copyright 2022 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+import {PluginType} from '../../metrics/types';
+import {TemporalScale} from '../histogram/histogram_types';
+import {Scale} from '../line_chart_v2/lib/public_types';
+
+export {Scale} from '../line_chart_v2/lib/public_types';
+
+export enum AxisDirection {
+  HORIZONTAL,
+  VERTICAL,
+}
+
+export enum Fob {
+  NONE,
+  START,
+  END,
+}
+
+export interface HistogramFobData {
+  type: PluginType.HISTOGRAMS;
```

Comment:
I find it suspicious that we would want or need to know the PluginType. My hope with the other suggestions is that we can eliminate this from the Fob data model and code.

Reply:
This type attribute is special and very useful. When we do a check like fobData.type === PluginType.HISTOGRAM the compiler now know that the fobData(which could be either HistogramFobData or ScalarFobData) is of the type HistogramFobData and will enforce that.

Reply:
I understand that you are using the field to help the compiler narrow the types. You use it to narrow FobCardData to one of ScalarFobData or HistogramFobData. It's a nice feature of the TypeScript compiler and we use it elsewhere in the code base.

I think, though, this is a misuse of type narrowing.

The end result in your controller code is that every function effectively looks like this:

```
someFunction() {
  if (this.fobCardData.type === PluginType.SCALARS) {
    // Do one thing for SCALARS.
    return;
  }

  // Do something completely different for HISTOGRAMS.
  return;
}
```

This is a anti-pattern. I think we can do better. My ideal is that SCALARS and HISTOGRAMS actually end up being similar enough that we don't need to treat then so differently. But if we do need to treat them so differently then there are better ways to do it, which we can discuss.